subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/GradleBuildExternalPluginsValidationSmokeTest.groovy

```diff
@@ -34,13 +35,36 @@ import org.gradle.test.preconditions.UnitTestPreconditions
 class GradleBuildExternalPluginsValidationSmokeTest extends AbstractGradleceptionSmokeTest implements WithPluginValidation, ValidationMessageChecker {
 
     def setup() {
-        allPlugins.projectPathToBuildDir = {
-            if (it == ':') {
-                'build'
+        ProjectBuildDirLocator locator = (String projectPath, TestFile projectRoot) -> {
+            if (projectPath == ':') {
+                projectRoot.file("build")
             } else {
-                "subprojects${it.split(':').join('/')}/build"
+                def projectDir = projectPath.split(':').join('/')
+
+                // Search in `subprojects` first.
+                def subprojectsPath = projectRoot.file("subprojects").file(projectDir).file("build")
+
+                if (subprojectsPath.exists()) {
+                    return subprojectsPath
+                }
+
+                // Otherwise, it might exist in a platform.
+                for (TestFile platformDir : projectRoot.file("platforms").listFiles()) {
+                    def platformPath = platformDir.file(projectDir).file("build")
+                    if (platformPath.exists()) {
+                        return platformPath
+                    }
+                }
+
+                throw new IllegalArgumentException("Cannot find build dir for project path '$projectPath'")
             }
         }
+
+        // Cache this since this gets called multiple times per project path.
+        Map<String, TestFile> locations = new HashMap<>()
+        allPlugins.projectPathToBuildDir = (String projectPath, TestFile projectRoot) -> {
```

Comment:
❔ Doesn't this algorithm need to be used in `WithPluginValidation` as well?

Reply:
Which algorithm? `allPlugins.projectPathToBuildDir` has a default value set in `WithPluginValidation`. This just overrides it to conform to the Gradle build's project structure

Reply:
I mean if another test other than this one implements the `WithPluginValidation` trait it will need to do this instead of using the default value.

Reply:
_do this_ as in override projectPathToBuildDir?

Reply:
Yes.  Am I misunderstanding?  Isn't the override in `GradleBuildExternalPluginsValidationSmokeTest` the universally correct algorithm?  Shouldn't that code just live in `WithPluginValidation` and there be no need for overriding?

Reply:
No. `AbstractPluginValidatingSmokeTest` also uses `WithPluginValidation` but applies the plugin to an empty project. It uses the default `projectPathToBuildDir` where projects follow the simple directory scheme defined by the default value of `projectPathToBuildDir`.